llama_index/query_engine/citation_query_engine.py

```diff
@@ -0,0 +1,299 @@
+    "which happens in the evening [1].\n"
+    "Now it's your turn. Below are several numbered sources of information:"
+    "\n------\n"
+    "{context_str}"
+    "\n------\n"
+    "Query: {query_str}\n"
+    "Answer: "
+)
+
+CITATION_REFINE_TEMPLATE = Prompt(
+    "Please provide an answer based only on the given sources. "
+    "When referencing information from a source, cite the appropriate source(s). "
+    "All answers should provide at least one source. "
+    "Do not use a source if it does not help answer a question. "
+    "For example:\n"
+    "Source 1:\n"
+    "The sky is red in the evening and blue in the morning.\n"
+    "Source 2:\n"
+    "Water is wet when the sky is red.\n"
+    "Query: When is water wet?\n"
+    "Answer: Water will be wet when the sky is red [2], "
+    "which happens in the evening [1].\n"
+    "Now it's your turn. "
+    "We have provided an existing answer: {existing_answer}"
+    "Below are several numbered sources of information. "
+    "Use them to refine the existing answer. "
+    "If the provided sources are not helpful, you will repeat the existing answer."
+    "\nBegin refining!"
+    "\n------\n"
+    "{context_msg}"
+    "\n------\n"
+    "Query: {query_str}\n"
+    "Answer: "
+)
+
+DEFAULT_SOURCE_CHUNK_SIZE = 512
+DEFAULT_SOURCE_CHUNK_OVERLAP = 20
+
+
+class CitaitonQueryEngine(BaseQueryEngine):
+    """Citation query engine.
+
+    Args:
+        retriever (BaseRetriever): A retriever object.
+        response_synthesizer (Optional[ResponseSynthesizer]):
+            A ResponseSynthesizer object.
+        source_node_size (int):
+            Size of source chunks, default=512. Useful for controlling
+            granularity of sources.
+        source_node_overlap (int): Overlap of source nodes, default=20.
+        text_splitter (Optional[TextSplitter]):
+            A text splitter for creating citation source nodes. Default is
+            a TokenTextSplitter.
+        callback_manager (Optional[CallbackManager]): A callback manager.
+    """
+
+    def __init__(
+        self,
+        retriever: BaseRetriever,
+        response_synthesizer: Optional[ResponseSynthesizer] = None,
+        source_chunk_size: int = DEFAULT_SOURCE_CHUNK_SIZE,
+        source_chunk_overlap: int = DEFAULT_SOURCE_CHUNK_OVERLAP,
+        text_splitter: Optional[TextSplitter] = None,
+        callback_manager: Optional[CallbackManager] = None,
+    ) -> None:
+        self.text_splitter = text_splitter or TokenTextSplitter(
+            chunk_size=source_chunk_size, chunk_overlap=source_chunk_overlap
+        )
+        self._retriever = retriever
+        self._response_synthesizer = (
+            response_synthesizer
+            or ResponseSynthesizer.from_args(callback_manager=callback_manager)
+        )
+
+        super().__init__(callback_manager)
+
+    @classmethod
+    def from_args(
+        cls,
+        index: BaseGPTIndex,
+        source_chunk_size: int = DEFAULT_SOURCE_CHUNK_SIZE,
```

Comment:
very nit: Is "citation_chunk_size" maybe a tiny bit more intuitive? On first read, I thought we need to set this to match the original chunk size of the documents.

Reply:
ah good idea